### DISCUSSION TITLE
Remove All Roles from specific role

### DISCUSSION THREAD

**Discussion by abanghendri** [spatie/laravel-permission#1789] [Q&A] [answered]
Hi, is there any function to remove all permission attached for role?

let say I want to flush all permission from 'member' role

thanks in advance

**Comment by drbyte** [comment, marked as answer]
```php
$role = Role::findByName('foo'); // or findById() or find()
$role->permissions->each(function ($role) {
  $role->delete();
});
```

**Reply by fredsal** [reply]
@drbyte is `$role->permissions->each(function ($role) { $role->delete(); });` deleting permission or relation role_has_permissions?

**Reply by drbyte** [reply]
Good point. It's deleting the permissions. 
Detaching can be done using the detach or sync examples posted below.

**Comment by fredsal** [comment]
Maybe better performance when you have a lot of permissions
```php
$role = Role::findByName('member'); // for get id by name
\DB::table(config('permission.table_names.role_has_permissions'))->where('role_id',$role->id)->delete();
$role->forgetCachedPermissions();
```
@drbyte is `$role->permissions->each(function ($role) { $role->delete(); });` deleting permission or relation `role_has_permissions`?

**Reply by abanghendri** [reply]
thanks @fredsal

**Reply by drbyte** [reply]
Note that if you use a direct `DB::` call to change the database content, you're risking corrupting your data if your foreign keys aren't firing (or if your db engine doesn't support foreign keys). 
And your cache will be out of sync until you create a new role or do a new role/permission assignment. The docs talk about how to rebuild the cache when you do such (unrecommended) manual data manipulation.

**Comment by erikn69** [comment]
>Sometimes it may be necessary to remove a role from a user. To remove a many-to-many relationship record, use the detach method. The detach method will delete the appropriate record out of the intermediate table; however, both models will remain in the database:

```php
$role = Role::findByName('member'); // get role by name

// Detach all permissions from the role...
$role->permissions()->detach();

// Rebuild the cache
$role->forgetCachedPermissions();
```
Also, maybe can work
```php
$role = Role::findByName('member');
$role->syncPermissions([]);
```
